doc/manpages/abctab2ps.1
.TH ABCTAB2PS 1 "18. May 2007"
.\"
.\"
.\"---------------------------------------------------------------------
.SH NAME
.\"---------------------------------------------------------------------
\fBabctab2ps\fP \- typesets music in abc format directly in postscript
.\"
.\"
.\"---------------------------------------------------------------------
.SH SYNOPSIS
.\"---------------------------------------------------------------------
abctab2ps [-1bcEhHIlnNopPx] 
[-a \fIalfa\fP] [-B \fIbarsperline\fP] 
[-d \fIdistance\fP] [-D \fIfmtdir\fP]
[-eCRST \fIselectors\fP] [-f \fIinfile\fP] 
[-F \fIformatfile\fP] [-g shrink|space|stretch|fill]
[-k \fInbars\fP] [-m \fImargin\fP] [-paper \fIformat\fP]
[-noauthor] [-notab] [-nofrenchtab] [-noitaliantab] [-nogermantab]
[-O \fIoutfile\fP] [-s \fIscale\fP]
[-t \fItransposition\fP]
[-tabsize \fIfontsize\fP] [-transposegchords]
[-v \fIverbosity\fP] [-V \fIvoices\fP]
[-w \fIwidth\fP] [-X \fIstrictness\fP]
[\fIinfile\fP[.abc]]
.\"
.\"
.\"---------------------------------------------------------------------
.SH DESCRIPTION
.\"---------------------------------------------------------------------
\fBabctab2ps\fP reads an input file containing music in abc format
and typesets it directly in PostScript.
For a description of the abc syntax, please see the separate user's guide.
.PP
Basically, the usage is \fIabctab2ps file1 file2 ...\fP 
where file1, file2 ... are the abc input files. This will write the
Postscript output to the file \fIOut.ps\fP; for a different filename
use flag -O. If no input file is specified, abctab2ps reads its input
from stdin so that you can pipe the output of some other abc tool into
abctab2ps.
.PP
Note that most flags (see OPTIONS) can be used in any sequence. 
However, the position does matter for flags e,f,C,S,R,T since
these determine how the following arguments are interpreted.
.PP
Flags can be contracted together if the meaning is clear, 
and a numerical argument for the last flag can be appended.
That is, "-b -s 0.7 -v 1" can be contracted to "-bs0.7 -v1".
.\"
.\"
.\"---------------------------------------------------------------------
.SH OPTIONS
.\"---------------------------------------------------------------------
.TP
.B -1
writes every tune on a separate page.
.TP
.B -a \fIalfa\fP
set the maximal amount of permitted shrinking to \fIalfa\fP,
where 0 <= \fIalfa\fP <= 1. A value of 0.0 means that no 
shrinking is allowed, and a value of 1.0 allows shrinking until 
the symbols are almost touching. In conjunction with -c, by default alfa 
is set to an intermediate value (displayed with "abctab2ps -c -h"). 
When -c is not used, maximal shrinking and stretching are allowed.
.TP
.B -b
break at all line ends, even if they end with the continuation 
symbol "\\".
.TP
.B -B \fIbarsperline\fP
try to typeset with \fIbarsperline\fP bars on each line.
.TP
.B -c
consider the input as one long line, ie., implicitly append 
the continuation symbol '\\' to every line of music.
When doing automatic line breaking with -c, the user can control the 
spacing of the symbols along the staff by specifying the "compression 
parameter" alfa with the -a flag.  
.TP
.B -d \fIdistance\fP
set staff distance to \fIdistance\fP (cm/in/pt).
.TP
.B -D \fIfmtdir\fP
look for format files in directory \fIfmtdir\fP.
.TP
.B -e \fIselector1 [selector2 ...]\fP
typesets only the tunes matching the xref numbers \fIselector1 
[selector2 ...]\fP. 
Optionally, the search can be done on other fields using in place of -e
the flags -R (seaches the rhythm field), -C (searches the composer field),
-S (searches the source field), -T (seaches the title field (default)).
.TP
.B -E
make EPSF (encapsulated postscript) output. Each tune is then put into 
a separate file with a correct bounding box and no page breaks.
.TP
.B -f \fIinfile\fP
sets the input file. Has the same effect as omitting this flag.
.TP
.B -F \fIformatfile\fP
read format from file \fIformatfile\fP.fmt. The directory can be set
with flag -D.
.TP
\fB-g\fP shrink|space|stretch|fill
sets the "glue mode".
The default mode is fill, which fills the staff.
This flag is useful when changing the layout parameters, 
to see what effect the changes have for each mode separately.
.TP
.B -h
quick help, equivalent to "abctab2ps" without any arguments.
This also shows the default settings for some parameters.
.TP
.B -H
show the default format parameters.
.TP
.B -I
write index file "Ind.ps".
.TP
.B -k \fInbars\fP
number every \fInbars\fP. To number every first bar per staff, set
\fInbars\fP to 0.
.TP
.B -l
print in landscape mode.
.TP
.B -m \fImargin\fP
sets the left margin to \fImargin\fP (cm/in/pt).
.TP
.B -maxs \fInsymbols\fP
obsolete option for different memory allocation.
No longer necessary because memory is now allocated dynamically.
.TP
.B -maxv \fInvoices\fP
obsolete option for different memory allocation.
No longer necessary because memory is now allocated dynamically.
.TP
.B -noauthor
suppresses the postscript DSC comment for the author (%%For).
Useful when abctab2ps is run automatically on a webserver or when
you do not want your name listed in the ouput file.
.TP
.B -notab
suppresses all tablature specific postscript routines and fonts
in the output file; useful for music only input, because the 
resulting output file is smaller. Basically, \fIabctab2ps\fP should
behave like \fIabc2ps\fP with this option.
.TP
.B -nofrenchtab
does not include the font for french tab into the output
file; useful if you only use a different tab, because the
resulting output file is smaller.
.TP
.B -noitaliantab
does not include the font for italian tab (and other number based
tabs) into the output file; useful if you only use a different tab,
because the resulting output file is smaller.
.TP
.B -nogermantab
does not include the font for german tab into the output
file; useful if you only use a different tab, because the
resulting output file is smaller.
.TP
.B -n
includes historical notes and other stuff at the bottom 
of each tune.
.TP
.B -N
write page numbers
.TP
.B -o
obsolete option that is ignored. In abc2ps, only with this option 
postscript output was generated.
.TP
.B -O \fIoutfile\fP
in PS (EPS) mode, the output is written to \fIoutfile\fP.ps
(\fIoutfile\fP001.eps). If the parameter to -O is '=', output 
for "foo.abc" is written to "foo.ps" ("foo001.eps").
.TP
.B -p
generates pretty output, with more whitespace between tunes,
larger fonts for titles, and larger music symbols. By default, 
the layout squeezes the tunes to reduce the number of pages.
.TP
.B -P
another pretty output format.
.TP
.B -paper \fIformat\fP
sets paper format. Typical values for \fIformat\fP are "a4" or "letter".
This command line option overrides the system-wide default papersize
given in \fI/etc/papersize\fP.
For a list of all possible values try an invalid value for \fIformat\fP;
this will also print the default papersize on your system.
.TP
.B -s \fIscale\fP
scales the music output by factor \fIscale\fP.
.TP
.B -t \fItransposition\fP
transpose all music. If \fItransposition\fP is a number, the music
is transposed up as many halftones; for downward transposition start 
the number with an underscore "_". If \fItransposition\fP is a note name, 
this note becomes the root.
.TP
.B -tabsize \fIfontsize\fP
sets the size of the tablature font TabFont to \fIfontsize\fP.
.TP
.B -transposegchords
transpose guitar chords too (in combination with option -t), 
in case you use guitar chords actually for chord letters
.TP
.B -v \fIverbosity\fP
sets the verbosity for output to the screen to \fIverbosity\fP.
Hereby -v0 gives very little, -v1,v2,v3.. show successively 
more information. Verbosity >= 10 is for debugging.
.TP
.B -V \fIvoices\fP
select specified voices, eg. "-V 1,4-5"
.TP
.B -w \fIwidth\fP
sets the width of the staff to \fIwidth\fP (cm/in/pt).
.TP
.B -x
includes the xref numbers in the tune title.
.TP
.B -X \fIstrictness\fP
set strictness for note spacing; 0 < \fIstrictness\fP < 1.
.\"
.\"
.\"---------------------------------------------------------------------
.SH ENVIRONMENT
.\"---------------------------------------------------------------------
.TP
.B ABCTABFONTS
Semicolon separated list of directories which abctab2ps searches
for tablature font files. If unset, fonts files are searched in
\fI/usr/share/abctab2ps\fP, \fI/usr/local/share/abctab2ps\fP and \fIfonts\fP
(in that order) on Unix/Win32/MacOs X and in \fI:fonts\fP on MacOs 8/9.
Please note that the separator is a semicolon ";" rather than a colon ":".
This is necessary because colons are part of path names on MacOs and Win32. 
.TP
.B PAPERSIZE
Sets the default paper format (\fIa4\fP or \fIletter\fP); overrides 
entry in \fI/etc/papersize\fP.
.TP
.B PAPERCONF
Full path to a file containing the default paper size (\fIa4\fP or 
\fIletter\fP).
If unset, \fI/etc/papersize\fP is used.
.\"
.\"
.\"---------------------------------------------------------------------
.SH FILES
.\"---------------------------------------------------------------------
.TP
.B /etc/papersize
Contains the name of the system-wide default paper size (\fIa4\fP or 
\fIletter\fP). A different file can be given with the environment variable
\fIPAPERCONF\fP.
.\"
.\"
.\"---------------------------------------------------------------------
.SH EXAMPLES
.\"---------------------------------------------------------------------
Typeset all tunes from mytunes.abc into Out.ps, choosing all line 
breaks automatically:
.PP
\	\fCabctab2ps mytunes.abc -c\fP
.PP
Squeeze notes together more and write the output to
mytunes.ps:
.PP
\	\fCabctab2ps mytunes.abc -O '=' -c -a0.9\fP
.PP
Equivalent commands to typeset tunes 1 to 10 in bla.abc and 11-20 
in fasel.abc:
.PP
\	\fCabctab2ps bla -e 1-10 -f fasel -e 11-20 \fP
.PP
\	\fCabctab2ps bla.abc 1-10 fasel.abc 11-20 \fP
.PP
\	\fCabctab2ps bla 1-10 fasel.abc 11-20 \fP
.PP
Typeset all tunes with the string 'House' in the title or with
xref numbers 10-12, in all abc files whose name starts with X,
including historical notes and xref numbers in the output, 
forcing a line break at continuation lines, with reduced size
of the symbols, putting one tune per page:
.PP
\	\fCabctab2ps X*.abc -e House 10,11,12 -nx -s0.9 -b1\fP
.\"
.\"
.\"---------------------------------------------------------------------
.SH BUGS
.\"---------------------------------------------------------------------
abctab2ps currently has a number of hard coded limits:
.IP -
Info fields must not exceed 300 characters
.IP -
file names must not exceed 100 characters
.IP -
the size of an output page is limited
.PP
If you encounter a limit, you currently need to change the hard coded
value in \fIabctab2ps.h\fP and recompile.
.\"
.\"
.\"---------------------------------------------------------------------
.SH AUTHOR
.\"---------------------------------------------------------------------
Christoph Dalitz <christoph(dot)dalitz[at]hs-niederrhein(dot)de>
.\"
.\"
.\"---------------------------------------------------------------------
.SH VERSION
.\"---------------------------------------------------------------------
This man page describes abctab2ps version 1.8.1 from May 2007.
This version is built on Michael Methfessel's abc2ps 1.3.3 from December 1998.
.\"
.\"
.\"---------------------------------------------------------------------
.SH COPYRIGHT
.\"---------------------------------------------------------------------
Copyright 1999-2007 by Christoph Dalitz
.PP
abctab2ps is supplied "as is" without any warranty. It
is free software and can be used, copied, modified and
distributed without fee under the terms of the GNU General 
Public License. 
.\"
.\"
.\"---------------------------------------------------------------------
.SH SEE ALSO
.\"---------------------------------------------------------------------
abcselect(1)


CHANGES
Change log of abctab2ps
=======================

Version 1.8.19, Mar 22 2022
---------------------------

 - new boolean option %%tab8underline to draw 8th course in frenchtab
   with underline and following courses with one slash less


Version 1.8.18, Mar 16 2022
---------------------------

 - new tab type "neapoltab" (thanks to Markus Lutz):
   like "spanishtab" or "guitartab", but starts with 1 instead of 0

 - diapasons in "germantab" drawn like in "frenchtab" (thanks to Markus Lutz)


Version 1.8.17, Sep 25 2020
---------------------------

 - fixed bug in bourdons for french5tab and french4tab

 - dotted rhythm in tab now adds flag on following chord to clarify rhythm


Version 1.8.16, May 25 2020
---------------------------

 - fixed bug that prevented tabdeco 'L' on top course


Version 1.8.15, Mar 11 2020
---------------------------

 - escaped percentage ("\%") recognized in %%center, %%right, and %%text lines

 - now error message when input filename too long


Version 1.8.14, Nov 05 2019
---------------------------

 - additional bourdon {15} supported


Version 1.8.13, May 25 2019
---------------------------

 - bourdons also supported in french4tab and french5tab
   (thanks to Markus Lutz)


Version 1.8.12, Jun 24 2015
---------------------------

 - ported to clang (OSX 10.10)

 - fixed PS error in dynamic marks


Version 1.8.11, Apr 26 2011
---------------------------

 - new boolean option %%tabfirstflag to draw flag
   only for rhythm changes or after bar line

 - when seventh course is used, additional bourdons moved down


Version 1.8.10, Mar 08 2011
---------------------------

 - new format parameter %%stafflinethickness

 - appogiaturas to chords now tied to first note in chord


Version 1.8.9, Dec 27 2009
--------------------------

 - new clef treble8up for soprano recorder


Version 1.8.8, Mar 14 2009
--------------------------

 - option -F no longer ignored in %! line


Version 1.8.7, Mar 10 2009
--------------------------

 - new format parameter %%tabgchordspace

 - size limit of 50 characters for words in text removed


Version 1.8.6, Aug 18 2008
--------------------------

 - corrected positioning of tabdecos for chords that only
   contain invisible notes (was broken since version 1.7.1)


Version 1.8.5, Aug 15 2008
--------------------------

 - new dynamic marks !sf! and !sfz!

 - separator line between systems in German lute tablature
   (can be suppressed with "%%tabgermansepline no")

 - last ledger line slightly wider for brevis and longa

 - all notes in a tune can be made stemless with %%nostems
   (useful for modern plainchant notation)

 - fixed an error in parsing special lyrics symbols '*' and '~'
   Escaping of '~' with "\~" now works again

 - frets 18-21 added in fonts itTimes and itBorrono
   (thanks to Martin Fick)


Version 1.8.4, Feb 21 2008
--------------------------

 - grace sequence before phrase mark "(" no longer lost during parsing

 - grace notes can now have length. The behaviour is as follows:

    o when no length specified, the old abc standard applies,
      i.e., single notes are accacciaturas (eighth note with stroke;
      the stroke can optionally be suppressed with %%nogracestroke)
      and more than one note means "sixteenth notes"

    o when a length is specified it applies to all grace notes
      in the same brace pair. When more than one length is given,
      the last length applies

 - fixed some possible causes for crashes when parsing grace sequence
   or decorations


Version 1.8.3, Dec 16 2007
--------------------------

 - new font deFraktur for German lute tablature (thanks to Michael Burzan)

 - new music decoration !wedge! (thanks to Michael Burzan)


Version 1.8.2, Aug 23 2007
--------------------------

 - corrected horizontal spacing errors for brevis and longa

 - longa stem now always at the right and downward

 - inline meter change now affects bar numbering even
   when "display=none" is used

 - fixed a potential crash when empty voice


Version 1.8.1, May 19 2007
--------------------------

 - author (full name of calling user) now written to PS file;
   in case of a PDF conversion, it will become the PDF author tag.
   Can be suppressed with command line option "-noauthor".

 - first bar number suppressed in tablature (now same as in music)

 - no longer double bar labels when label on last bar in line.
   Label on last bar in line is always moved to beginning of next
   line. To suppress this, put the label on an invisible bar
   before the last bar in the line.

 - no longer floating exception when default note length cannot be
   determined due to inproperly closed header. abctab2ps now stops
   and reports stopping point in abc file instead.

 - trailing comma in tab chords no longer causes erroneous postscript


Version 1.8.0, Apr 23 2007
--------------------------

 - printing of metronome marks (Q:...) for abc2midi can be
   suppressed in postscript output with "%%printmetronome no"

 - prelimary support for German lute tablature (see example waissel.abc)
   (still lacks a nice Fraktur font)


Version 1.7.1, Apr 17 2007
--------------------------

 - epm description file abctab2ps.list corrected
   => userguide now correctly installed in Linux RPM and DEB

 - new format parameter %%nogracestroke to suppress stroke through
   single eighth gracenotes

 - tablature: y (invisible fret symbol) ignored in positioning decos 
   that apply to an entire chord (right hand fingering etc.)


Version 1.7.0, Jul 31 2006
--------------------------

 - tenuto strokes in tablature chords with !ten(! and !ten)!

 - notehead size in historic style reduced

 - fixed a memory error in music.cpp:draw_sysbars()


Version 1.6.8, Mar 18 2006
--------------------------

 - segfault fixed when empty V: field

 - long lyrics word under melisma no longer causes uneven spacing


Version 1.6.7, Mar 16 2006
--------------------------

 - abc2midi voice parameter "octave" ignored


src/abctab2ps.cpp
/*
 *  abctab2ps: a program to typeset tunes in abc format using PostScript
 *  Copyright (C) 1996 - 1998  Michael Methfessel
 *  Copyright (C) 1999 - 2019  Christoph Dalitz
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 *  The author can be contacted as follows:
 *  
 *  Christoph Dalitz
 *  <christoph (dot) dalitz (at) hs-niederrhein (dot) de>
 *
 */

#include <stdlib.h>
#include <string.h>
#include <signal.h>

#include "abctab2ps.h"
#include "subs.h"
#include "format.h"
#include "util.h"
#include "parse.h"
#include "music.h"
#include "oscompat.h"


/* ----- definitions of global variables ----- */

int db=DEBUG_LV;                  /* debug control */
int maxSyms,maxVc;                /* for malloc */
int allocSyms,allocVc;
struct ISTRUCT info,default_info; /* information fields */
char lvoiceid[STRLINFO];          /* string from last V: line */
int  nvoice,mvoice;               /* number of voices defd, nonempty */
int  ivc;                         /* current voice */
int  ivc0;                        /* top nonempty voice */
int ixpfree;                      /* first free element in xp array */
struct METERSTR default_meter;    /* data to specify the meter */
struct KEYSTR default_key;        /* data to specify the key */
                          /* things to alloc: */
struct SYMBOL  *sym;              /* symbol list */
struct SYMBOL  **symv;            /* symbols for voices */
struct XPOS    *xp;               /* shared horizontal positions */
struct VCESPEC *voice;            /* characteristics of a voice */
struct SYMBOL  **sym_st;          /* symbols a staff start */
int            *nsym_st; 

int halftones;                    /* number of halftones to transpose by */

float f0p,f5p,f1p,f0x,f5x,f1x;            /*   mapping fct */
float lnnp,bnnp,fnnp,lnnx,bnnx,fnnx;      /*   note-note spacing */
float lbnp,bbnp,rbnp,lbnx,bbnx,rbnx;      /*   bar-note spacing */
float lnbp,bnbp,rnbp,lnbx,bnbx,rnbx;      /*   note-bar spacing */


char wpool[NWPOOL];            /* pool for vocal strings */
int nwpool,nwline;             /* globals to handle wpool */

struct SYMBOL zsym;            /* symbol containing zeros */

struct FORMAT sfmt;                    /* format after initialization */
struct FORMAT dfmt;                    /* format at start of tune */
struct FORMAT cfmt;                    /* current format for output */

char fontnames[50][STRLFMT];           /* list of needed fonts */
int  nfontnames;

//char txt[MAXNTEXT][MAXWLEN];           /* for output of text */
//int  ntxt; 
StringVector words_of_text;            /* for output of text */


char vcselstr[STRLFILE];          /* string for voice selection */
char mbf[501];                 /* mini-buffer for one line */
char buf[BUFFSZ];              /* output buffer.. should hold one tune */
int nbuf;                      /* number of bytes buffered */
float bposy;                   /* current position in buffered data */
int   ln_num;                  /* number of lines in buffer */
float ln_pos[BUFFLN];          /* vertical positions of buffered lines */
int   ln_buf[BUFFLN];          /* buffer location of buffered lines */
int   use_buffer;              /* 1 if lines are being accumulated */

char text [NTEXT][STRLINFO];   /* pool for history, words, etc. lines */
int text_type[NTEXT];          /* type of each text line */
int ntext;                     /* number of text lines */
char page_init[201];           /* initialization string after page break */
int do_mode;                   /* control whether to do index or output */
char escseq[81];               /* escape sequence string */
int linenum;                   /* current line number in input file */
int tunenum;                   /* number of current tune */
int tnum1,tnum2;
int numtitle;                  /* how many titles were read */
int mline;                     /* number music lines in current tune */
int nsym;                      /* number of symbols in line */
int nsym0;                     /* nsym at start of parsing a line */
int pagenum;                   /* current page in output file */
int writenum;                  /* calls to write_buffer for each one tune */
int xrefnum;                   /* xref number of current tune */
int do_meter, do_indent;       /* how to start next block */

int index_pagenum;             /* for index file */
float index_posx, index_posy;
int index_initialized;

GchordList prep_gchlst;          /* guitar chords for preparsing */
Deco prep_deco;                  /* decorations for preparsing */
int bagpipe;                     /* switch for HP mode */
int within_tune, within_block;   /* where we are in the file */
int do_this_tune;                /* are we typesetting the current one ? */
float posx,posy;                 /* overall scale, position on page */
int barinit;                     /* carryover bar number between parts */

char *p, *p0;                    /* global pointers for parsing music line */

int word,slur;                   /* variables used for parsing... */
int last_note,last_real_note;
int pplet,qplet,rplet;
int carryover;                   /* for interpreting > and < chars */
int ntinext,tinext[MAXHD];       /* for chord ties */

struct ENDINGS ending[20];       /* where to draw endings */           
int num_ending;                  /* number of endings to draw */
int mes1,mes2;                   /* to count measures in an ending */

int slur1[20],slur2[20];         /* needed for drawing slurs */
int overfull;                    /* flag if staff overfull */
int do_words;                    /* flag if staff has words under it */

int vb, verbose;                 /* verbosity, global and within tune */
int in_page=0;

                                 /* switches modified by flags: */
int gmode;                         /* switch for glue treatment */
int include_xrefs;                 /* to include xref numbers in title */
int one_per_page;                  /* new page for each tune ? */
int pagenumbers;                   /* write page numbers ? */
int write_history;                 /* write history and notes ? */
int help_me;                       /* need help ? */
int select_all;                    /* select all tunes ? */
int epsf;                          /* for EPSF postscript output */
int choose_outname;                /* 1 names outfile w. title/fnam */
int break_continues;               /* ignore continuations ? */
int search_field0;                 /* default search field */
int pretty;                        /* for pretty but sprawling layout */
int bars_per_line;                 /* bars for auto linebreaking */
int continue_lines;                /* flag to continue all lines */
int landscape;                     /* flag for landscape output */
int barnums;                       /* interval for bar numbers */
int make_index;                    /* write index file ? */
int notab;                         /* do not process tablature */
int transposegchords;              /* transpose gchords */
float alfa_c;                      /* max compression allowed */
float scalefac;                    /* scale factor for symbol size */
float lmargin;                     /* left margin */
float swidth;                      /* staff width */
float staffsep,dstaffsep;          /* staff separation */
float strict1,strict2;             /* 1stave, mstave strictness */
char transpose[21];                /* target key for transposition */
struct PAPERSIZE* paper;           /* paper size (a4, letter) */
int noauthor;                      /* suppress PS author tag (%%For) */


float alfa_last,beta_last;         /* for last short short line.. */

char in_file[MAXINF][STRLFILE];  /* list of input file names */
int  ninf;                       /* number of input file names */
FILE *fin;                       /* for input file */

char outf[STRLFILE];             /* output file name */
char outfnam[STRLFILE];          /* internal file name for open/close */
char styf[STRLFILE];             /* layout style file name */
char styd[STRLFILE];             /* layout style directory */
char infostr[STRLFILE];          /* title string in PS file */

int  file_open;                  /* for output file */
int  file_initialized;           /* for output file */
FILE *fout,*findex;              /* for output file */
int nepsf;                       /* counter for epsf output files */

char sel_str[MAXINF][STRLFILE];  /* list of selector strings */
int  s_field[MAXINF];            /* type of selection for each file */
int  psel[MAXINF];               /* pointers from files to selectors */

int temp_switch;

/* signal handler for premature termination */
void sighnd_term(int sig)
{
  close_output_file();
  exit(130);
}


/* ----- start of main ------ */
int main(int argc, char **argv)
{
  char ext[41];
  char xref_str[STRLFILE], pat[30][STRLFILE];
  int isel,j,npat,search_field,retcode,rc;

  /* cleanup on premature termination */
  if ((signal(SIGINT,sighnd_term)==SIG_ERR) ||
      (signal(SIGTERM,sighnd_term)==SIG_ERR)) {
    wng("could not install signal handler for SIGTERM and SIGINT","");
  }

  /* ----- set default options and parse arguments ----- */

  maxSyms = allocSyms = 800;
  maxVc   = allocVc = 3;

  init_ops (1);
  retcode=parse_args (argc, argv);
  if (retcode) exit (1);
  if (do_mode==DO_OUTPUT) {
    printf("This is abctab2ps, version %s.%s (%s)\n", VERSION, REVISION, VDATE);
  }

  alloc_structs ();

  /* ----- set the page format ----- */
  nfontnames=0;
  if (!set_page_format()) exit (3);
  if (help_me==2) {
    print_format(cfmt);
    exit (0);
  }
  
  /* ----- help printout  ----- */
  if (help_me==1) {
    write_help ();
    exit (0);
  }
  
  if (ninf>0) isel=psel[ninf-1];
  else isel=psel[0];
  search_field0=s_field[isel];   /* default for interactive mode */
  if (epsf) cutext(outf);
  
  /* ----- initialize ----- */
  zero_sym();
  pagenum=0;
  tunenum=tnum1=tnum2=0;
  verbose=0;
  file_open=file_initialized=0;
  nepsf=0;
  bposy=0;
  posx=cfmt.leftmargin;
  posy=cfmt.pageheight-cfmt.topmargin; 

  strcpy(page_init, "");

  if ((do_mode == DO_OUTPUT) && make_index) open_index_file (INDEXFILE);

  /* ----- loop over files in list ---- */
  if (ninf==0) printf ("++++ No input files, read from stdin\n");
  for (j=0;j<=ninf;j++) {
    if (ninf==0) {
      /* no input file specified: open stdin */
      fin = stdin;
      strcpy(in_file[0],"stdin");
    } else {
      /* process list of input files */
      if (j==ninf) {
        break;
      } else {
        getext (in_file[j],ext);
        /*  skip .ps and .eps files */
        if ((!strcmp(ext,"ps"))||(!strcmp(ext,"eps"))) continue;

        if ((fin = fopen (in_file[j],"r")) == NULL) {      
          if (!strcmp(ext,"")) strext (in_file[j],in_file[j],"abc",1);
          if ((fin = fopen (in_file[j],"r")) == NULL) { 
            printf ("++++ Cannot open input file: %s\n", in_file[j]);
            continue;
          }
        }
      }
    }

    isel=psel[j];
    search_field=s_field[isel];
    npat=rehash_selectors (sel_str[isel], xref_str, pat);
    dfmt=sfmt;
    strnzcpy(infostr, in_file[j], STRLFILE);

    if (do_mode==DO_INDEX) {
      printf ("%s:\n", in_file[j]);
      do_index (fin,xref_str,npat,pat,select_all,search_field);
    }
      
    else {
      if (!epsf) {
        strext (outf, outf, "ps", 1);
        if (choose_outname) strext (outf, in_file[j], "ps", 1);
        open_output_file(outf,in_file[j]);
      }
      printf ("%s: ", in_file[j]); 
      if (vb>=3) printf ("\n"); 
      process_file (fin,fout,xref_str,npat,pat,select_all,search_field);
      printf ("\n");
    }
      
  }
  
  if (do_mode==DO_INDEX)
    printf ("Selected %d title%s of %d\n", tnum1, tnum1==1?"":"s", tnum2);
  
  if ((do_mode == DO_OUTPUT) && make_index) close_index_file ();
  rc = close_output_file ();

  if ((do_mode == DO_OUTPUT) && rc)
    return 1;
  else
    return 0;
}


